platform/mbed_wait_api_no_rtos.c

```diff
@@ -31,8 +31,8 @@ void wait_ms(int ms) {
 
 void wait_us(int us) {
     const ticker_data_t *const ticker = get_us_ticker_data();
-    uint32_t start = ticker_read(ticker);
-    while ((ticker_read(ticker) - start) < (uint32_t)us);
+    us_timestamp_t start = ticker_read_us(ticker);
+    while ((ticker_read_us(ticker) - start) < (us_timestamp_t)us);
```

Comment:
What is the problem with the way this was initially handled? Could you mock up an example showing where this would have failed (like making a stub version of ticker_read returning values before and after the overflow point)?

Casting `us` to a `us_timestamp_t` might have unintended side effects for negative numbers. I would expect passing negative values to `wait_us` to either trap with an error or return immediately. In the case of this cast it will block indefinitely.

Reply:
I have check it, everything works fine with 32bit version.
(https://github.com/maciejbocianski/mbed-os/tree/mbed_wait_us_test - branch with test, 
test command `mbed test -t GCC_ARM -m k64f -n tests-mbed_platform-wait -vv --profile develop`)

My bad, forgot that when subtracting unsigned integers modular arithmetic (also known as "wrap around" behavior) applies.  In the result `0 - 4294967285` gives `11`

```
[1530170876.01][CONN][RXD] start: 4294967285
[1530170876.04][CONN][RXD] current: 4294967286  diff:  1
[1530170876.07][CONN][RXD] current: 4294967287  diff:  2
[1530170876.11][CONN][RXD] current: 4294967288  diff:  3
[1530170876.14][CONN][RXD] current: 4294967289  diff:  4
[1530170876.17][CONN][RXD] current: 4294967290  diff:  5
[1530170876.21][CONN][RXD] current: 4294967291  diff:  6
[1530170876.22][CONN][RXD] current: 4294967292  diff:  7
[1530170876.27][CONN][RXD] current: 4294967293  diff:  8
[1530170876.31][CONN][RXD] current: 4294967294  diff:  9
[1530170876.33][CONN][RXD] current: 4294967295  diff: 10
[1530170876.36][CONN][RXD] current:          0  diff: 11
[1530170876.39][CONN][RXD] current:          1  diff: 12
[1530170876.43][CONN][RXD] current:          2  diff: 13
[1530170876.46][CONN][RXD] current:          3  diff: 14
[1530170876.49][CONN][RXD] current:          4  diff: 15
[1530170876.53][CONN][RXD] current:          5  diff: 16
[1530170876.56][CONN][RXD] current:          6  diff: 17
[1530170876.59][CONN][RXD] current:          7  diff: 18
[1530170876.62][CONN][RXD] current:          8  diff: 19
```